src/library/features/crates/cratefeature.cpp

```diff
@@ -45,13 +45,35 @@ CrateFeature::CrateFeature(UserSettingsPointer pConfig,
           m_cratesIcon(":/images/library/ic_library_crates.png"),
           m_lockedCrateIcon(":/images/library/ic_library_locked.png"),
           m_pTrackCollection(pTrackCollection),
-          m_pCrateTableModel(nullptr) {
+          m_pCrateTableModel(nullptr),
+          m_crateHierarchy(pTrackCollection) {
+
 
     initActions();
 
     // construct child model
-    m_childModel.setRootItem(std::make_unique<TreeItem>(this));
-    rebuildChildModel();
+    // m_childModel.setRootItem(std::make_unique<TreeItem>(this));
+    // rebuildChildModel();
+
+    // if closure does not have the same number of crates as the crates table
+    // this means that the user just started mixxx with nested crates for the
+    // first time, so we have to fill the closure table with (self,self,0)
+    if (!m_crateHierarchy.closureIsValid()) {
+        QMessageBox::warning(
+            nullptr,
+            tr("Nested Crates"),
+            tr("Crates now support hierarchical structure."
+               "All your crates have been converted to Level 1 crates"));
+
+        m_crateHierarchy.resetClosure();
+        m_crateHierarchy.initClosure();
+        m_crateHierarchy.resetPath();
+        m_crateHierarchy.generateAllPaths();
```

Comment:
Do you anticipate needing these 4 functions in any other context? If not, it may make sense to make them private members of CrateHierarchy and create a public wrapper function (maybe called CrateHierarchy::initializeHierarchy or something like that) around them to call here. I think this exposes too much of the implementation of the database storage to the GUI class.